### ISSUE TITLE
SqlitePclRawStorageEngine RunInTransaction blocks the thread

### ISSUE THREAD

**Issue by vguzmanp** [couchbase/couchbase-lite-net#736]
There is a Task.Wait in [this code](https://github.com/couchbase/couchbase-lite-net/blob/master/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqlitePclRawStorageEngine.cs#L460) that deadlocks the main thread.
You can see the explanation of the deadlock [here](http://blog.stephencleary.com/2012/07/dont-block-on-async-code.html)

**Comment by borrrden** [comment]
I've seen that page before and I don't see why it explains this situation.  The scheduler sends all its items to a privately created thread and so while there is the potential for a deadlock if you wait on a task from inside runintransaction I have yet to run into it.   Perhaps you could show what you are doing and say which platform?

**Comment by vguzmanp** [comment]
I you check our fork (https://github.com/fluendo/couchbase-lite-net/tree/CBLite-deadlock) you will find a project called Fluendo.Tests. This tests our modifications, but these modifications don't affect that section of the code.

Executing this test on OSX using Mono 4.2.1 triggers the deadlock. Removing the whole Task from the code - making the code synchronous - makes it disappear.

**Comment by borrrden** [comment]
Can you send the details of the deadlock (stack trace of the threads at the time it occurs)?

**Comment by vguzmanp** [comment]
I ran the test with debug activated, and paused the execution after a minute or so. It's paused on the `Wait`:

Stack trace:

```
﻿﻿System.Threading.Monitor.Monitor_wait() in 
System.Threading.Monitor.Wait(object obj, int millisecondsTimeout) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/mcs/class/corlib/System.Threading/Monitor.cs:148
System.Threading.ManualResetEventSlim.Wait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/ManualResetEventSlim.cs:669
System.Threading.Tasks.Task.SpinThenBlockingWait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:3356
System.Threading.Tasks.Task.InternalWait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:3295
System.Threading.Tasks.Task.Wait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:3201
System.Threading.Tasks.Task.Wait() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:3083
Couchbase.Lite.Storage.SystemSQLite.SqlitePCLRawStorageEngine.RunInTransaction(Couchbase.Lite.RunInTransactionDelegate block) in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqlitePclRawStorageEngine.cs:453
Couchbase.Lite.Storage.SystemSQLite.SqliteCouchStore.RunInTransaction(Couchbase.Lite.RunInTransactionDelegate block) in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqliteCouchStore.cs:953
Couchbase.Lite.Storage.SystemSQLite.SqliteCouchStore.OptimizeSQLIndexes() in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqliteCouchStore.cs:246
Couchbase.Lite.Storage.SystemSQLite.SqliteCouchStore.Open() in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqliteCouchStore.cs:514
Couchbase.Lite.Storage.SystemSQLite.SqliteCouchStore.Open(string directory, Couchbase.Lite.Manager manager, bool readOnly) in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqliteCouchStore.cs:865
Couchbase.Lite.Database.OpenWithOptions(Couchbase.Lite.DatabaseOptions options) in /Users/vguzman/Development/couchbase-lite-net/src/Couchbase.Lite.Shared/Database.cs:2041
Couchbase.Lite.Manager.OpenDatabase(string name, Couchbase.Lite.DatabaseOptions options) in /Users/vguzman/Development/couchbase-lite-net/src/Couchbase.Lite.Shared/Manager.cs:399
Couchbase.Lite.Manager.GetDatabase(string name) in /Users/vguzman/Development/couchbase-lite-net/src/Couchbase.Lite.Shared/Manager.cs:416
FluendoUnitTests.SQLSearch.TestCase() in /Users/vguzman/Development/couchbase-lite-net/src/Fluendo.Tests/Test.cs:16
```

Threads:

```
Id 			Name						Location
1 										0x14 in System.Net.Sockets.Socket_Receive_Internal
2 Parent process watcher 				0x1 in MonoDevelopProcessHost.WatchParentProcess.AnonymousMethod__0 at Users/builder/data/lanes/3055/f3c0d982/source/monodevelop/main/src/tools/mdhost/src/mdhost.cs:154,4
3 										0xF in System.Net.Sockets.Socket.Accept_internal
4 										0xD in System.Threading.Thread.SleepInternal
5 										0x1A in System.Threading.Thread.JoinInternal
6 										0xF in System.Net.Sockets.Socket.Accept_internal
7 										0x1C in System.Threading.WaitHandle.WaitOne_internal
8 EventPumpThread 						OxE in System.Threading.Monitor.Monitor_wait
9 TestRunnerThread 						OxE in System.Threading.Monitor.Monitor_wait
10 Finalizer
12 <Thread Pool>
13 <Thread Pool>
14 Database Thread 						OxE in System.Threading.Monitor.Monitor_wait
```

**Comment by borrrden** [comment]
To qualify as a deadlock there needs to be at least two threads waiting for each other.  That means I need at least two stack traces.  Which thread is that trace from?  TestRunnerThread?  What is the trace on Database Thread?

**Comment by zgramana** [comment]
One caveat to that would be when a TaskScheduler of MaxConcurrency == 1 runs a Task that pushes another Task onto that same work queue and then waits for it to complete.

**Comment by zgramana** [comment]
(Unless you allow RunInline)

**Comment by borrrden** [comment]
Actually the scheduler is smart enough to realize when things are queued from the same work queue since everything is on one thread.  In those cases it doesn't queue the item and executes it out of order directly since the only reason for that to happen is for a "dependent" task.

**Comment by zgramana** [comment]
FWIW, that mechanism works only if you properly implement [TryExecuteTaskInline] (https://msdn.microsoft.com/en-us/library/system.threading.tasks.taskscheduler.tryexecutetaskinline(v=vs.110).aspx)

**Comment by borrrden** [comment]
Oh, I wasn't referring to the built in logic.  I have extra logic that intercepts tasks before they are queued :).  I also attempted to implement `TryExecuteTaskInline` just in case.

**Comment by zgramana** [comment]
\o/

**Comment by zgramana** [comment]
One scenario that popped up and was very hard to find in Java-Core was a Future that indirectly waited on itself via `Result`. Same problem, but much harder to detect.

**Comment by vguzmanp** [comment]
The stack trace I pasted was from the TestRunnerThread. The other ones are:

EventPumpThread:
```
System.Threading.Monitor.Monitor_wait() in 
System.Threading.Monitor.Wait(object obj, int millisecondsTimeout) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/mcs/class/corlib/System.Threading/Monitor.cs:148
System.Threading.Monitor.Wait(object obj) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/mcs/class/corlib/System.Threading/Monitor.cs:134
NUnit.Core.EventQueue.Dequeue(bool blockWhenEmpty) in 
NUnit.Core.EventPump.PumpThreadProc() in 
System.Threading.ThreadHelper.ThreadStart_Context(System.Threading.ThreadHelper state) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/thread.cs:68
System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Threading.ThreadHelper state, bool preserveSyncCtx) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:581
System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Threading.ThreadHelper state, bool preserveSyncCtx) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:530
System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Threading.ThreadHelper state) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:519
System.Threading.ThreadHelper.ThreadStart() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/thread.cs:105
```

DatabaseThread:

```
System.Threading.Monitor.Monitor_wait() in 
System.Threading.Monitor.Wait(object obj, int millisecondsTimeout) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/mcs/class/corlib/System.Threading/Monitor.cs:148
System.Threading.ManualResetEventSlim.Wait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/ManualResetEventSlim.cs:669
System.Threading.Tasks.Task.SpinThenBlockingWait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:3356
System.Threading.Tasks.Task.InternalWait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:3295
System.Threading.Tasks.Task<bool>.GetResultCore(bool waitCompletionNotification) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Future.cs:562
System.Threading.Tasks.Task<bool>.get_Result() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Future.cs:535
Couchbase.Lite.Storage.SystemSQLite.SqlitePCLRawStorageEngine.BeginTransaction() in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqlitePclRawStorageEngine.cs:314
Couchbase.Lite.Storage.SystemSQLite.SqlitePCLRawStorageEngine.RunInTransaction.AnonymousMethod__0() in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqlitePclRawStorageEngine.cs:417
System.Threading.Tasks.Task.InnerInvoke() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:2901
System.Threading.Tasks.Task.Execute() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:2523
System.Threading.Tasks.Task.ExecutionContextCallback( obj) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:2887
System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback,  state, bool preserveSyncCtx) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:581
System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback,  state, bool preserveSyncCtx) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:530
System.Threading.Tasks.Task.ExecuteWithThreadLocal( currentTaskSlot) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:2848
System.Threading.Tasks.Task.ExecuteEntry(bool bPreventDoubleExecution) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:2781
System.Threading.Tasks.TaskScheduler.TryExecuteTask( task) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/TaskScheduler.cs:455
Couchbase.Lite.Util.SingleThreadScheduler.Drain() in /Users/vguzman/Development/couchbase-lite-net/src/Couchbase.Lite.Shared/Util/SingleThreadScheduler.cs:102
Couchbase.Lite.Util.SingleThreadScheduler.Run() in /Users/vguzman/Development/couchbase-lite-net/src/Couchbase.Lite.Shared/Util/SingleThreadScheduler.cs:78
System.Threading.ThreadHelper.ThreadStart_Context(System.Threading.ThreadHelper state) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/thread.cs:68
System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Threading.ThreadHelper state, bool preserveSyncCtx) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:581
System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Threading.ThreadHelper state, bool preserveSyncCtx) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:530
System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Threading.ThreadHelper state) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:519
System.Threading.ThreadHelper.ThreadStart() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/thread.cs:105
```

And now I have another waiting thread `12 <Thread Pool>`:

```
System.Threading.Monitor.Monitor_wait() in 
System.Threading.Monitor.Wait(object obj, int millisecondsTimeout) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/mcs/class/corlib/System.Threading/Monitor.cs:148
System.Threading.ManualResetEventSlim.Wait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/ManualResetEventSlim.cs:669
System.Threading.Tasks.Task.SpinThenBlockingWait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:3356
System.Threading.Tasks.Task.InternalWait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:3295
System.Threading.Tasks.Task<Couchbase.Lite.Cursor>.GetResultCore(bool waitCompletionNotification) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Future.cs:562
System.Threading.Tasks.Task<Couchbase.Lite.Cursor>.get_Result() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Future.cs:535
Couchbase.Lite.Storage.SystemSQLite.SqlitePCLRawStorageEngine.IntransactionRawQuery(string sql, object[] paramArgs) in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqlitePclRawStorageEngine.cs:511
Couchbase.Lite.Storage.SystemSQLite.SqliteCouchStore.TryQuery(System.Func<Couchbase.Lite.Cursor,bool> action, bool readUncommit, string sqlQuery, object[] args) in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqliteCouchStore.cs:349
Couchbase.Lite.Storage.SystemSQLite.SqliteCouchStore.QueryOrDefault<long?>(System.Func<Couchbase.Lite.Cursor,long?> action, bool readUncommit, long? defaultVal, string sqlQuery, object[] args) in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqliteCouchStore.cs:377
Couchbase.Lite.Storage.SystemSQLite.SqliteCouchStore.NextDocumentExpiry() in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqliteCouchStore.cs:1132
Couchbase.Lite.Database.SchedulePurgeExpired(System.TimeSpan delay) in /Users/vguzman/Development/couchbase-lite-net/src/Couchbase.Lite.Shared/Database.cs:2110
Couchbase.Lite.Database.PurgeExpired(object state) in /Users/vguzman/Development/couchbase-lite-net/src/Couchbase.Lite.Shared/Database.cs:2160
System.Threading.Timer.Scheduler.TimerCB(System.Threading.Timer o) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/mcs/class/corlib/System.Threading/Timer.cs:327
System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/threadpool.cs:1264
System.Threading.ThreadPoolWorkQueue.Dispatch() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/threadpool.cs:859
System.Threading._ThreadPoolWaitCallback.PerformWaitCallback() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/threadpool.cs:1196
``

**Comment by borrrden** [comment]
The presence of `IntransactionRawQuery` in the stack trace tells me that either your assemblies aren't up to date (Couchbase.Lite.Storage.SystemSQLite) or your fork is based off of an older commit.  I factored out that call among others as part of some improvements to avoid situations like this.

**Comment by vguzmanp** [comment]
Our fork is based off `release/1.3.1`. Is that change included in the release?

**Comment by borrrden** [comment]
Yes, in that case I'd make sure that you actually got all the updated assemblies into your output.  The beginning of your logs should show commits of relevant assemblies and when they were built.  Note that you need to build and copy your storage engine assembly separately.

**Comment by vguzmanp** [comment]
Can you point me to some of the commits that make this change?

**Comment by borrrden** [comment]
https://github.com/couchbase/couchbase-lite-net/commit/8293147815427d864c19930c6028264613d1dc01 This is the commit where I removed the method I mentioned

**Comment by borrrden** [comment]
Oh!  It looks like that it a master-only change.  However, this commit will be in 1.3.2.

**Comment by vguzmanp** [comment]
Testing with release/1.3.2 (commit 0812d750bcf1442b94f73dfb162ef81737deb8f7) I still reproduce this deadlock. I have 3 threads in `0xE in System.Threading.Monitor.Monitor_wait`:

Database Thread:
```
System.Threading.Monitor.Monitor_wait() in 
System.Threading.Monitor.Wait(object obj, int millisecondsTimeout) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/mcs/class/corlib/System.Threading/Monitor.cs:148
System.Threading.ManualResetEventSlim.Wait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/ManualResetEventSlim.cs:669
System.Threading.Tasks.Task.SpinThenBlockingWait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:3356
System.Threading.Tasks.Task.InternalWait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:3295
System.Threading.Tasks.Task<bool>.GetResultCore(bool waitCompletionNotification) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Future.cs:562
System.Threading.Tasks.Task<bool>.get_Result() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Future.cs:535
Couchbase.Lite.Storage.SystemSQLite.SqlitePCLRawStorageEngine.BeginTransaction() in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqlitePclRawStorageEngine.cs:321
Couchbase.Lite.Storage.SystemSQLite.SqlitePCLRawStorageEngine.RunInTransaction.AnonymousMethod__0() in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqlitePclRawStorageEngine.cs:424
System.Threading.Tasks.Task.InnerInvoke() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:2901
System.Threading.Tasks.Task.Execute() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:2523
System.Threading.Tasks.Task.ExecutionContextCallback( obj) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:2887
System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback,  state, bool preserveSyncCtx) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:581
System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback,  state, bool preserveSyncCtx) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:530
System.Threading.Tasks.Task.ExecuteWithThreadLocal( currentTaskSlot) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:2848
System.Threading.Tasks.Task.ExecuteEntry(bool bPreventDoubleExecution) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:2781
System.Threading.Tasks.TaskScheduler.TryExecuteTask( task) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/TaskScheduler.cs:455
Couchbase.Lite.Util.SingleThreadScheduler.Drain() in /Users/vguzman/Development/couchbase-lite-net/src/Couchbase.Lite.Shared/Util/SingleThreadScheduler.cs:108
Couchbase.Lite.Util.SingleThreadScheduler.Run() in /Users/vguzman/Development/couchbase-lite-net/src/Couchbase.Lite.Shared/Util/SingleThreadScheduler.cs:84
System.Threading.ThreadHelper.ThreadStart_Context(System.Threading.ThreadHelper state) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/thread.cs:68
System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Threading.ThreadHelper state, bool preserveSyncCtx) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:581
System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Threading.ThreadHelper state, bool preserveSyncCtx) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:530
System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Threading.ThreadHelper state) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:519
System.Threading.ThreadHelper.ThreadStart() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/thread.cs:105
```

TestRunnerThread:

```
System.Threading.Monitor.Monitor_wait() in 
System.Threading.Monitor.Wait(object obj, int millisecondsTimeout) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/mcs/class/corlib/System.Threading/Monitor.cs:148
System.Threading.ManualResetEventSlim.Wait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/ManualResetEventSlim.cs:669
System.Threading.Tasks.Task.SpinThenBlockingWait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:3356
System.Threading.Tasks.Task.InternalWait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:3295
System.Threading.Tasks.Task.Wait(int millisecondsTimeout,  cancellationToken) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:3201
System.Threading.Tasks.Task.Wait() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/Tasks/Task.cs:3083
Couchbase.Lite.Storage.SystemSQLite.SqlitePCLRawStorageEngine.RunInTransaction(Couchbase.Lite.RunInTransactionDelegate block) in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqlitePclRawStorageEngine.cs:460
Couchbase.Lite.Storage.SystemSQLite.SqliteCouchStore.RunInTransaction(Couchbase.Lite.RunInTransactionDelegate block) in /Users/vguzman/Development/couchbase-lite-net/src/StorageEngines/SQLiteCommon/storage.sqlite.common/src/SqliteCouchStore.cs:949
Couchbase.Lite.Database.RunInTransaction(Couchbase.Lite.RunInTransactionDelegate transactionDelegate) in /Users/vguzman/Development/couchbase-lite-net/src/Couchbase.Lite.Shared/Database.cs:732
FluendoUnitTests.SQLSearch.TestCase() in /Users/vguzman/Development/couchbase-lite-net/src/Fluendo.Tests/Test.cs:20
```

EventPumpThread:

```
System.Threading.Monitor.Monitor_wait() in 
System.Threading.Monitor.Wait(object obj, int millisecondsTimeout) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/mcs/class/corlib/System.Threading/Monitor.cs:148
System.Threading.Monitor.Wait(object obj) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/mcs/class/corlib/System.Threading/Monitor.cs:134
NUnit.Core.EventQueue.Dequeue(bool blockWhenEmpty) in 
NUnit.Core.EventPump.PumpThreadProc() in 
System.Threading.ThreadHelper.ThreadStart_Context(System.Threading.ThreadHelper state) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/thread.cs:68
System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Threading.ThreadHelper state, bool preserveSyncCtx) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:581
System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Threading.ThreadHelper state, bool preserveSyncCtx) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:530
System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Threading.ThreadHelper state) in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/executioncontext.cs:519
System.Threading.ThreadHelper.ThreadStart() in /Users/vguzman/oneplay-build/sources-devel/darwin_x86_64/mono-4.2.1/external/referencesource/mscorlib/system/threading/thread.cs:105
```

**Comment by vguzmanp** [comment]
This happens with Mono 4.2.1. Testing with latest Mono (4.6.0) I cannot reproduce it.

**Comment by borrrden** [comment]
Mono 4.2.1 is not really a supportable platform for us because in 4.4 there were some sweeping (and breaking) changes with regards to TLS support.  In general each release is tested against the latest release of Mono.  It also looks like you built this 4.2.1 yourself, did you make any changes to Mono itself?

Just FYI, EventPumpThread is specific to NUnit and will not pass through any Couchbase code.  It just handles some NUnit bookkeeping stuff.

**Comment by djpongh** [comment]
Given that Mono 4.2.1 isn't supportable by us and this works in Mono 4.6.0, I'm closing.
